File: apps/scala/design-pattern/test/BuilderTest.scala
package designpattern.builder

class BuilderSuite extends munit.FunSuite:

  test("Builder でコンピュータを構築する") {
    val computer = ComputerBuilder()
      .setCpu("Intel i7")
      .setMemory(16)
      .setStorage(512)
      .build()

    assertEquals(computer.cpu, "Intel i7")
    assertEquals(computer.memory, 16)
    assertEquals(computer.storage, 512)
    assertEquals(computer.gpu, None)
  }

  test("全オプションを指定して構築する") {
    val computer = ComputerBuilder()
      .setCpu("AMD Ryzen 9")
      .setMemory(64)
      .setStorage(2000)
      .setGpu("RTX 4080")
      .setDisplay("32インチ")
      .build()

    assertEquals(computer.gpu, Some("RTX 4080"))
    assertEquals(computer.display, Some("32インチ"))
  }

  test("ゲーミング PC の定型構成") {
    val gaming = ComputerDirector.gamingComputer

    assertEquals(gaming.cpu, "Intel i9")
    assertEquals(gaming.memory, 32)
    assertEquals(gaming.gpu, Some("RTX 4090"))
  }

  test("オフィス PC の定型構成") {
    val office = ComputerDirector.officeComputer

    assertEquals(office.cpu, "Intel i5")
    assertEquals(office.gpu, None)
  }

  test("describe でスペックを表示する") {
    val laptop = ComputerDirector.laptopComputer
    val desc   = laptop.describe

    assert(desc.contains("CPU: Apple M3"))
    assert(desc.contains("ディスプレイ: 14インチ Retina"))
  }

  test("case class の copy でイミュータブルに変更する") {
    val base     = ComputerDirector.officeComputer
    val upgraded = base.copy(memory = 32, gpu = Some("RTX 3060"))

    assertEquals(base.memory, 16)
    assertEquals(upgraded.memory, 32)
  }

File: apps/clojure/design-pattern/test/design_pattern/builder_test.clj
(ns design-pattern.builder-test
  (:require [clojure.test :refer :all]
            [design-pattern.builder :refer :all]))

(deftest basic-builder-test
  (testing "スレッディングマクロでコンピュータを構築する"
    (let [pc (-> (new-computer)
                 (with-cpu "Intel Core i7")
                 (with-ram 16)
                 (with-storage 512)
                 build-computer)]
      (is (= "Intel Core i7" (:cpu pc)))
      (is (= 16 (:ram-gb pc)))
      (is (= 512 (:storage-gb pc)))
      (is (:built pc)))))

(deftest validation-fails-test
  (testing "必須フィールドが欠けている場合は例外を投げる"
    (is (thrown? clojure.lang.ExceptionInfo
                (-> (new-computer)
                    (with-cpu "Intel")
                    build-computer)))))

(deftest gaming-computer-test
  (testing "ゲーミング PC プリセット"
    (let [pc (gaming-computer)]
      (is (= "NVIDIA RTX 4090" (:gpu pc)))
      (is (= 32 (:ram-gb pc)))
      (is (:built pc)))))

(deftest office-computer-test
  (testing "オフィス PC プリセット"
    (let [pc (office-computer)]
      (is (= "Intel Core i5" (:cpu pc)))
      (is (nil? (:gpu pc)))
      (is (:built pc)))))

(deftest build-from-specs-test
  (testing "仕様マップからコンピュータを構築する"
    (let [specs {:cpu "AMD Ryzen 9" :ram-gb 64 :storage-gb 4000 :gpu "NVIDIA RTX 4080"}
          pc    (build-computer (build-from-specs specs))]
      (is (= "AMD Ryzen 9" (:cpu pc)))
      (is (= 64 (:ram-gb pc)))
      (is (:built pc)))))
